### PR TITLE
many: add Client-User-Agent to "SnapAction" install API call

### PR DESCRIPTION
**Pull request by mvo5**
Also set the client-user-agent when doing an install operation. This PR is RFC because it cheats a bit, the client-user-agent is just set when the "SnapAction()" API call is performed but not on the download. We could do the later but it means we need to store the client-user-agent in the task/change somehow. Given that the action api is probably enough I did not do that.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -1173,6 +1174,7 @@ func postSnap(c *Command, r *http.Request, user *auth.UserState) Response {
 	if user != nil {
 		inst.userID = user.ID
 	}
+	inst.ctx = store.WithClientUserAgent(r.Context(), r)
```
Comment (pedronis) [inline daemon/api.go]: as we discussed the last time around, I think we should move the WithClientUserAgent to Command.ServeHTTP itself

Comment (mvo5) [reply daemon/api.go]: Indeed, I fixed this now.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 2]
```diff
@@ -447,7 +447,7 @@ func (s *servicectlSuite) TestQueuedCommandsSingleLane(c *C) {
 	s.st.Lock()
 
 	chg := s.st.NewChange("install change", "install change")
-	ts, err := snapstate.Install(s.st, "one", &snapstate.RevisionOptions{Revision: snap.R(1)}, 0, snapstate.Flags{})
+	ts, err := snapstate.Install(context.TODO(), s.st, "one", &snapstate.RevisionOptions{Revision: snap.R(1)}, 0, snapstate.Flags{})
```
Comment (chipaca) [inline overlord/hookstate/ctlcmd/services_test.go]: this isn't really a TODO though :-) or is it?

File: overlord/snapstate/handlers.go [hunk 3]
```diff
@@ -253,7 +254,7 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
 	}
 
 	// not installed, nor queued for install -> install it
-	ts, err := Install(st, snapName, &RevisionOptions{Channel: channel}, userID, Flags{RequireTypeBase: requireTypeBase})
+	ts, err := Install(context.TODO(), st, snapName, &RevisionOptions{Channel: channel}, userID, Flags{RequireTypeBase: requireTypeBase})
```
Comment (chipaca) [inline overlord/snapstate/handlers.go]: how much bigger does this get if you make these take the context as well?

File: overlord/snapstate/storehelpers.go [hunk 4]
```diff
@@ -118,7 +118,7 @@ func installInfo(st *state.State, name string, revOpts *RevisionOptions, userID
 
 	theStore := Store(st, deviceCtx)
 	st.Unlock() // calls to the store should be done without holding the state lock
-	res, err := theStore.SnapAction(context.TODO(), curSnaps, []*store.SnapAction{action}, user, opts)
+	res, err := theStore.SnapAction(ctx, curSnaps, []*store.SnapAction{action}, user, opts)
```
Comment (chipaca) [inline overlord/snapstate/storehelpers.go]: woo

File: overlord/snapstate/snapstate_test.go [hunk 5]
```diff
@@ -555,16 +555,16 @@ func (s *snapmgrTestSuite) TestInstallDevModeConfinementFiltering(c *C) {
 
 	// if a snap is devmode, you can't install it without --devmode
 	opts := &snapstate.RevisionOptions{Channel: "channel-for-devmode"}
-	_, err := snapstate.Install(s.state, "some-snap", opts, s.user.ID, snapstate.Flags{})
+	_, err := snapstate.Install(context.TODO(), s.state, "some-snap", opts, s.user.ID, snapstate.Flags{})
```
Comment (chipaca) [inline overlord/snapstate/snapstate_test.go]: unless the plan is to have some sort of context you can pass to these from the test, none of these are TODOs, and having so many of them make finding the ones we actually want a lot harder.

File: daemon/api_test.go [hunk 6]
```diff
@@ -3640,6 +3641,32 @@ func (s *apiSuite) testInstall(c *check.C, forcedDevmode bool, flags snapstate.F
...
+	req, err := http.NewRequest("POST", "/v2/snaps/some-snap", &buf)
+	req.RemoteAddr = "pid=100;uid=0;socket=;"
+	c.Assert(err, check.IsNil)
+	req.Header.Add("User-Agent", "some-agent/1.0")
+
+	s.vars = map[string]string{"name": "some-snap"}
+	rec := httptest.NewRecorder()
+	snapCmd.ServeHTTP(rec, req)
+	c.Assert(rec.Code, check.Equals, 202)
+	c.Check(store.ClientUserAgent(s.ctx), check.Equals, "some-agent/1.0")
```
Comment (pedronis) [inline daemon/api_test.go]: we might want c.Assert(s.ctx, check.NotNil) before this one

File: overlord/devicestate/devicestate.go [hunk 7]
```diff
@@ -351,7 +352,7 @@ func remodelTasks(st *state.State, current, new *asserts.Model, deviceCtx snapst
 		_, err := snapstate.CurrentInfo(st, snapName)
 		// If the snap is not installed we need to install it now.
 		if _, ok := err.(*snap.NotInstalledError); ok {
-			ts, err := snapstateInstallWithDeviceContext(st, snapName, nil, userID, snapstate.Flags{Required: true}, deviceCtx)
+			ts, err := snapstateInstallWithDeviceContext(context.TODO(), st, snapName, nil, userID, snapstate.Flags{Required: true}, deviceCtx)
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go]: remodelTasks should take the context from outside I think

File: overlord/hookstate/ctlcmd/services_test.go (hunk 2 above)
Comment (mvo5) [reply overlord/hookstate/ctlcmd/services_test.go]: Not really, I was following the exiting UpdateMany() whichhas a TODO context but that is wrong as well so fixing them all now :)

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (mvo5) [reply overlord/snapstate/handlers.go]: I can do this in a followup. Note that right now we only use the context when we create the tasks (in the snapaction install API call to the store). So adding those will not add value right away.

File: overlord/devicestate/devicestate.go (hunk 7 above)
Comment (mvo5) [reply overlord/devicestate/devicestate.go]: We can do this now (http://paste.ubuntu.com/p/57CgcQ3chp/) and I'm happy to push this. But unless I miss something it won't add much at this point. The snapstate.Remodel() function does not take a context yet and it does not currently need one. So we might add the context once we have a use-case. But either way is fine with me.

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (chipaca) [reply overlord/snapstate/handlers.go]: yeah, leave it for later

**Comment by anonymouse64** [comment]
commit 7579e301a3c8ac576ed6cff70366d19991c71a03 here seems to have introduced an issue whereby if you don't remove the `github.com/gorilla/context` dependency and build snapd with the script at the bottom you get an error like this:

```bash
$ SNAP_REEXEC=0 ./snap-exe install hello-world
error: cannot install "": cannot install snap with empty name
```

This seems to be the reason for one of the next commits https://github.com/snapcore/snapd/pull/7100/commits/be4fc4d117c255cadd697a93f9f94e49c708f2c3, but still it's a bit confusing to track down this error. I'm not sure if there's anything actionable to resolve this other than to teach others like myself to _always_ run `get-deps.sh` before building snapd.

The script I used with git bisect to find this issue (after having disable re-exec by putting `SNAP_REEXEC=0` in the systemctl override file for `snapd.service`):

```bash
./mkversion.sh
go build -o snapd ./cmd/snapd
go build -o snap-exe ./cmd/snap
pushd cmd
autoreconf -i -f
make clean
sudo make hack install
popd
sudo systemctl stop snapd
sudo cp snapd /usr/lib/snapd/snapd
sudo systemctl restart snapd
SNAP_REEXEC=0 ./snap-exe install hello-world
SNAP_REEXEC=0 ./snap-exe remove hello-world
```

**Pull request merged**